
Save New Duplicate & Edit Just Text
diff --git a/arch/arm/dts/sun55i-a523.dtsi b/arch/arm/dts/sun55i-a523.dtsi
index 5dbfebfcbde..31aa97443fa 100644
--- a/arch/arm/dts/sun55i-a523.dtsi
+++ b/arch/arm/dts/sun55i-a523.dtsi
@@ -3,10 +3,16 @@
 
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/clock/sun6i-rtc.h>
+#include <dt-bindings/clock/sun55iw3-mcu-ccu.h>
 #include <dt-bindings/clock/sun55i-a523-ccu.h>
 #include <dt-bindings/clock/sun55i-a523-r-ccu.h>
+#include <dt-bindings/reset/sun55iw3-mcu-ccu.h>
 #include <dt-bindings/reset/sun55i-a523-ccu.h>
 #include <dt-bindings/reset/sun55i-a523-r-ccu.h>
+#include <dt-bindings/power/allwinner,sun55i-a523-ppu.h>
+#include <dt-bindings/power/allwinner,sun55i-a523-pck600.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/phy/phy.h>
 
 / {
 	interrupt-parent = <&gic>;
@@ -106,6 +112,61 @@
 		#size-cells = <1>;
 		ranges = <0x0 0x0 0x0 0x40000000>;
 
+		usbc1: usbc1@11 {
+			device_type = "usbc1";
+			reg = <0x11 0x1000>;
+			usb_regulator_io = "nocare";
+			usb_wakeup_suspend = <0>;
+			status = "disabled";
+		};
+
+		usbc2:usbc2@12 {
+			device_type = "usbc2";
+			compatible = "allwinner,sunxi-plat-dwc3";
+			reg = <0x0 0x12 0x0 0x1000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			ranges;
+			aw,hcgen2-phygen1-quirk;
+			status = "disabled";
+
+			xhci2: xhci2-controller@4d00000 {
+				compatible = "snps,dwc3";
+				reg = <0x0 0x04d00000 0x0 0x100000>;
+				interrupts = <GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
+				dr_mode = "otg"; // dr_mode option: host, peripheral, otg
+				clocks = <&ccu CLK_USB3_MBUS_GATE>, <&ccu CLK_USB3_REF>, <&ccu CLK_USB2_REF>, <&ccu CLK_USB3_SUSPEND>;
+				clock-names = "bus_clk", "ref_clk3", "ref_clk2", "suspend";
+				resets = <&mcu_ccu RST_BUS_3>;
+				reset-names = "hci";
+				maximum-speed = "super-speed";
+				phy_type = "utmi";
+				snps,dis_enblslpm_quirk;
+				snps,dis-u1-entry-quirk;
+				snps,dis-u2-entry-quirk;
+				snps,dis_u3_susphy_quirk;
+				snps,dis_u2_susphy_quirk;
+				phys = <&u2phy>, <&combophy PHY_TYPE_USB3>;
+				phy-names = "usb2-phy", "usb3-phy";
+				status = "disabled";
+			};
+		};
+
+		gpu: gpu@1800000 {
+			compatible = "allwinner,sun55i-a523-mali",
+				     "arm,mali-valhall-jm";
+			reg = <0x1800000 0x10000>;
+			interrupts = <GIC_SPI 117 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 118 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 119 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "job", "mmu", "gpu";
+			clocks = <&ccu CLK_GPU>, <&ccu CLK_BUS_GPU>;
+			clock-names = "core", "bus";
+			power-domains = <&pck600 PD_GPU>;
+			resets = <&ccu RST_BUS_GPU>;
+			status = "disabled";
+		};
+
 		pio: pinctrl@2000000 {
 			compatible = "allwinner,sun55i-a523-pinctrl";
 			reg = <0x2000000 0x800>;
@@ -144,92 +205,77 @@
 			};
 
 			mmc2_pins: mmc2-pins {
-				pins = "PC1" ,"PC5", "PC6", "PC8", "PC9",
-				       "PC10", "PC11", "PC13", "PC14", "PC15",
-				       "PC16";
+				pins = "PC0", "PC1" ,"PC5", "PC6", "PC8",
+				       "PC9", "PC10", "PC11", "PC13", "PC14",
+				       "PC15", "PC16";
 				allwinner,pinmux = <3>;
 				function = "mmc2";
 				drive-strength = <30>;
 				bias-pull-up;
 			};
 
+			rgmii0_pins: rgmii0-pins {
+				pins = "PH0", "PH1", "PH2", "PH3", "PH4",
+				       "PH5", "PH6", "PH7", "PH9", "PH10",
+				       "PH14", "PH15", "PH16", "PH17", "PH18";
+				allwinner,pinmux = <5>;
+				function = "gmac0";
+				drive-strength = <40>;
+				bias-disable;
+			};
+
+
+			rgmii1_pins: rgmii1-pins {
+				pins = "PJ0", "PJ1", "PJ2", "PJ3", "PJ4",
+				       "PJ5", "PJ6", "PJ7", "PJ8", "PJ9",
+				       "PJ11", "PJ12", "PJ13", "PJ14", "PJ15";
+				allwinner,pinmux = <5>;
+				function = "gmac1";
+				drive-strength = <40>;
+				bias-disable;
+			};
+
 			uart0_pb_pins: uart0-pb-pins {
 				pins = "PB9", "PB10";
 				allwinner,pinmux = <2>;
 				function = "uart0";
 			};
+
+			/omit-if-no-ref/
+			uart1_pins: uart1-pins {
+				pins = "PG6", "PG7";
+				function = "uart1";
+				allwinner,pinmux = <2>;
+			};
+
+			/omit-if-no-ref/
+			uart1_rts_cts_pins: uart1-rts-cts-pins {
+				pins = "PG8", "PG9";
+				function = "uart1";
+				allwinner,pinmux = <2>;
+			};
 		};
 
-		ccu: clock@2001000 {
+		ccu: clock-controller@2001000 {
 			compatible = "allwinner,sun55i-a523-ccu";
 			reg = <0x02001000 0x1000>;
-			clocks = <&osc24M>, <&rtc CLK_OSC32K>, <&rtc CLK_IOSC>;
-			clock-names = "hosc", "losc", "iosc";
+			clocks = <&osc24M>, <&rtc CLK_OSC32K>,
+				 <&rtc CLK_IOSC>, <&rtc CLK_OSC32K_FANOUT>;
+			clock-names = "hosc", "losc",
+				      "iosc", "losc-fanout";
 			#clock-cells = <1>;
 			#reset-cells = <1>;
 		};
 
-		mmc0: mmc@4020000 {
-			compatible = "allwinner,sun55i-a523-mmc",
-				     "allwinner,sun20i-d1-mmc";
-			reg = <0x04020000 0x1000>;
-			clocks = <&ccu CLK_BUS_MMC0>, <&ccu CLK_MMC0>;
-			clock-names = "ahb", "mmc";
-			resets = <&ccu RST_BUS_MMC0>;
-			reset-names = "ahb";
-			interrupts = <GIC_SPI 40 IRQ_TYPE_LEVEL_HIGH>;
-			pinctrl-names = "default";
-			pinctrl-0 = <&mmc0_pins>;
-			status = "disabled";
-
-			max-frequency = <150000000>;
-			cap-sd-highspeed;
-			cap-mmc-highspeed;
-			cap-sdio-irq;
-			#address-cells = <1>;
-			#size-cells = <0>;
-		};
-
-		mmc1: mmc@4021000 {
-			compatible = "allwinner,sun55i-a523-mmc",
-				     "allwinner,sun20i-d1-mmc";
-			reg = <0x04021000 0x1000>;
-			clocks = <&ccu CLK_BUS_MMC1>, <&ccu CLK_MMC1>;
-			clock-names = "ahb", "mmc";
-			resets = <&ccu RST_BUS_MMC1>;
-			reset-names = "ahb";
-			interrupts = <GIC_SPI 41 IRQ_TYPE_LEVEL_HIGH>;
-			pinctrl-names = "default";
-			pinctrl-0 = <&mmc1_pins>;
-			status = "disabled";
-
-			max-frequency = <150000000>;
-			cap-sd-highspeed;
-			cap-mmc-highspeed;
-			cap-sdio-irq;
-			#address-cells = <1>;
-			#size-cells = <0>;
-		};
-
-		mmc2: mmc@4022000 {
-			compatible = "allwinner,sun55i-a523-mmc",
-				     "allwinner,sun20i-d1-mmc";
-			reg = <0x04022000 0x1000>;
-			clocks = <&ccu CLK_BUS_MMC2>, <&ccu CLK_MMC2>;
-			clock-names = "ahb", "mmc";
-			resets = <&ccu RST_BUS_MMC2>;
-			reset-names = "ahb";
-			interrupts = <GIC_SPI 42 IRQ_TYPE_LEVEL_HIGH>;
-			pinctrl-names = "default";
-			pinctrl-0 = <&mmc2_pins>;
-			status = "disabled";
-
-			max-frequency = <150000000>;
-			cap-sd-highspeed;
-			cap-mmc-highspeed;
-			cap-sdio-irq;
-			#address-cells = <1>;
-			#size-cells = <0>;
+		mmu_aw: iommu@2010000 {
+			compatible = "allwinner,iommu-v15-sun55iw3";
+			reg = <0x0 0x02010000 0x0 0x1000>;
+			interrupts = <GIC_SPI 65 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "iommu-irq";
+			clocks = <&ccu CLK_IOMMU>;
+			clock-names = "iommu";
+			/* clock-frequency = <24000000>; */
+			#iommu-cells = <2>;
 		};
 
 		wdt: watchdog@2050000 {
@@ -407,6 +453,28 @@
 			#size-cells = <0>;
 		};
 
+		syscon: syscon@3000000 {
+			compatible = "allwinner,sun55i-a523-system-control",
+				     "allwinner,sun50i-a64-system-control";
+			reg = <0x03000000 0x1000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			ranges;
+		};
+
+		dma:dma-controller@3002000 {
+			compatible = "allwinner,dma-v105";
+			reg = <0x03002000 0x1000>;
+			interrupts = <GIC_SPI 50 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&ccu CLK_BUS_DMA>, <&ccu CLK_MBUS_DMA>;
+			clock-names = "bus", "mbus";
+			dma-channels = <8>;
+			dma-requests = <54>;
+			resets = <&ccu RST_BUS_DMA>;
+			#dma-cells = <1>;
+			status = "okay";
+		};
+
 		gic: interrupt-controller@3400000 {
 			compatible = "arm,gic-v3";
 			#address-cells = <1>;
@@ -428,6 +496,69 @@
 			};
 		};
 
+		mmc0: mmc@4020000 {
+			compatible = "allwinner,sun55i-a523-mmc",
+				     "allwinner,sun20i-d1-mmc";
+			reg = <0x04020000 0x1000>;
+			clocks = <&ccu CLK_BUS_MMC0>, <&ccu CLK_MMC0>;
+			clock-names = "ahb", "mmc";
+			resets = <&ccu RST_BUS_MMC0>;
+			reset-names = "ahb";
+			interrupts = <GIC_SPI 40 IRQ_TYPE_LEVEL_HIGH>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&mmc0_pins>;
+			status = "disabled";
+
+			max-frequency = <150000000>;
+			cap-sd-highspeed;
+			cap-mmc-highspeed;
+			cap-sdio-irq;
+			#address-cells = <1>;
+			#size-cells = <0>;
+		};
+
+		mmc1: mmc@4021000 {
+			compatible = "allwinner,sun55i-a523-mmc",
+				     "allwinner,sun20i-d1-mmc";
+			reg = <0x04021000 0x1000>;
+			clocks = <&ccu CLK_BUS_MMC1>, <&ccu CLK_MMC1>;
+			clock-names = "ahb", "mmc";
+			resets = <&ccu RST_BUS_MMC1>;
+			reset-names = "ahb";
+			interrupts = <GIC_SPI 41 IRQ_TYPE_LEVEL_HIGH>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&mmc1_pins>;
+			status = "disabled";
+
+			max-frequency = <150000000>;
+			cap-sd-highspeed;
+			cap-mmc-highspeed;
+			cap-sdio-irq;
+			#address-cells = <1>;
+			#size-cells = <0>;
+		};
+
+		mmc2: mmc@4022000 {
+			compatible = "allwinner,sun55i-a523-mmc",
+				     "allwinner,sun20i-d1-mmc";
+			reg = <0x04022000 0x1000>;
+			clocks = <&ccu CLK_BUS_MMC2>, <&ccu CLK_MMC2>;
+			clock-names = "ahb", "mmc";
+			resets = <&ccu RST_BUS_MMC2>;
+			reset-names = "ahb";
+			interrupts = <GIC_SPI 42 IRQ_TYPE_LEVEL_HIGH>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&mmc2_pins>;
+			status = "disabled";
+
+			max-frequency = <150000000>;
+			cap-sd-highspeed;
+			cap-mmc-highspeed;
+			cap-sdio-irq;
+			#address-cells = <1>;
+			#size-cells = <0>;
+		};
+
 		usb_otg: usb@4100000 {
 			compatible = "allwinner,sun55i-a523-musb",
 				     "allwinner,sun8i-a33-musb";
@@ -519,6 +650,142 @@
 			status = "disabled";
 		};
 
+		gmac0: ethernet@4500000 {
+			compatible = "allwinner,sun55i-a523-emac0",
+				     "allwinner,sun50i-a64-emac";
+			reg = <0x04500000 0x10000>;
+			clocks = <&ccu CLK_BUS_EMAC0>;
+			clock-names = "stmmaceth";
+			resets = <&ccu RST_BUS_EMAC0>;
+			reset-names = "stmmaceth";
+			interrupts = <GIC_SPI 46 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "macirq";
+			pinctrl-names = "default";
+			pinctrl-0 = <&rgmii0_pins>;
+			syscon = <&syscon>;
+			status = "disabled";
+
+			mdio0: mdio {
+				compatible = "snps,dwmac-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+			};
+		};
+
+		gmac1: ethernet@4510000 {
+			compatible = "allwinner,sun55i-a523-gmac200",
+				     "snps,dwmac-4.20a";
+			reg = <0x04510000 0x10000>;
+			clocks = <&ccu CLK_BUS_EMAC1>, <&ccu CLK_MBUS_EMAC1>;
+			clock-names = "stmmaceth", "mbus";
+			resets = <&ccu RST_BUS_EMAC1>;
+			reset-names = "stmmaceth";
+			interrupts = <GIC_SPI 47 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "macirq";
+			pinctrl-names = "default";
+			pinctrl-0 = <&rgmii1_pins>;
+			power-domains = <&pck600 PD_VO1>;
+			syscon = <&syscon>;
+			snps,fixed-burst;
+			snps,axi-config = <&gmac1_stmmac_axi_setup>;
+			snps,mtl-rx-config = <&gmac1_mtl_rx_setup>;
+			snps,mtl-tx-config = <&gmac1_mtl_tx_setup>;
+			status = "disabled";
+
+			mdio1: mdio {
+				compatible = "snps,dwmac-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+			};
+
+			gmac1_mtl_rx_setup: rx-queues-config {
+				snps,rx-queues-to-use = <1>;
+
+				queue0 {};
+			};
+
+			gmac1_stmmac_axi_setup: stmmac-axi-config {
+				snps,wr_osr_lmt = <0xf>;
+				snps,rd_osr_lmt = <0xf>;
+				snps,blen = <256 128 64 32 16 8 4>;
+			};
+
+			gmac1_mtl_tx_setup: tx_queues-config {
+				snps,tx-queues-to-use = <1>;
+
+				queue0 {};
+			};
+		};
+
+		pcie: pcie@4800000 {
+			compatible = "allwinner,sunxi-pcie-v210-rc";
+			#address-cells = <3>;
+			#size-cells = <2>;
+			bus-range = <0x0 0xff>;
+			reg = <0x04800000 0x480000>;
+			reg-names = "dbi";
+			device_type = "pci";
+			ranges = <0x00000800 0 0x20000000 0x20000000 0 0x01000000
+				  0x81000000 0 0x21000000 0x21000000 0 0x01000000
+				  0x82000000 0 0x22000000 0x22000000 0 0x0e000000>;
+			num-lanes = <1>;
+			phys = <&combophy PHY_TYPE_PCIE>;
+			phy-names = "pcie-phy";
+			interrupts = <GIC_SPI 107 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 106 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 98  IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 99  IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 100 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 101 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 102 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 103 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 104 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 105 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "msi", "sii", "edma-w0", "edma-w1", "edma-w2", "edma-w3",
+					  "edma-r0", "edma-r1", "edma-r2", "edma-r3";
+			#interrupt-cells = <1>;
+			interrupt-controller;
+			num-edma = <4>;
+			max-link-speed = <2>;
+			num-ib-windows = <8>;
+			num-ob-windows = <8>;
+			linux,pci-domain = <0>;
+			power-domains = <&pck600 PD_PCIE>;
+			clocks = <&osc24M>, <&ccu CLK_PCIE_AUX>;
+			clock-names = "hosc", "pclk_aux";
+			status = "disabled";
+		};
+
+		u2phy: phy@4e00000 {
+			compatible = "allwinner,sunxi-plat-phy";
+			reg = <0x0 0x04e00000 0x0 0x800>; /* Application Registers */
+			#phy-cells = <0>;
+			status = "disabled";
+		};
+
+		combophy: phy@4f00000 {
+			compatible = "allwinner,inno-combphy";
+			reg = <0x04f00000 0x80000>, /* Sub-System Application Registers */
+			      <0x04f80000 0x80000>; /* Combo INNO PHY Registers */
+			reg-names = "phy-ctl", "phy-clk";
+			power-domains = <&pck600 PD_PCIE>;
+			phy_refclk_sel = <0>; /* 0:internal clk; 1:external clk */
+			clocks = <&ccu CLK_USB3_REF>, <&ccu CLK_PLL_PERIPH0_200M>;
+			clock-names = "phyclk_ref","refclk_par";
+			resets = <&ccu RST_BUS_PCIE_USB3>;
+			reset-names = "phy_rst";
+			#phy-cells = <1>;
+			status = "disabled";
+		};
+
+		ppu: power-controller@7001400 {
+			compatible = "allwinner,sun55i-a523-ppu";
+			reg = <0x07001400 0x400>;
+			clocks = <&r_ccu CLK_BUS_R_PPU1>;
+			resets = <&r_ccu RST_BUS_R_PPU1>;
+			#power-domain-cells = <1>;
+		};
+
 		r_ccu: clock-controller@7010000 {
 			compatible = "allwinner,sun55i-a523-r-ccu";
 			reg = <0x7010000 0x250>;
@@ -565,6 +832,14 @@
 			};
 		};
 
+		pck600: power-controller@7060000 {
+			compatible = "allwinner,sun55i-a523-pck-600";
+			reg = <0x07060000 0x8000>;
+			clocks = <&r_ccu CLK_BUS_R_PPU0>;
+			resets = <&r_ccu RST_BUS_R_PPU0>;
+			#power-domain-cells = <1>;
+		};
+
 		r_i2c0: i2c@7081400 {
 			compatible = "allwinner,sun55i-a523-i2c",
 				     "allwinner,sun8i-v536-i2c",
@@ -592,5 +867,26 @@
 			clock-names = "bus", "hosc", "ahb";
 			#clock-cells = <1>;
 		};
+
+		mcu_ccu: mcu_ccu@7102000 {
+			compatible = "allwinner,sun55iw3-mcu-ccu";
+			reg = <0x07102000 0x165>;
+			#clock-cells = <1>;
+			#reset-cells = <1>;
+		};
+
+		dma1:dma1-controller@7121000 {
+			compatible = "allwinner,dma-v104";
+			reg = <0x7121000 0x1000>;
+			interrupts = <GIC_SPI 197 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&mcu_ccu CLK_BUS_MCU_DMA>, <&mcu_ccu CLK_BUS_MCU_DMA_MBUS>, <&mcu_ccu CLK_BUS_MCU_MBUS>;
+			clock-names = "bus", "mbus", "mcu-mbus";
+			dma-channels = <8>;
+			dma-requests = <15>;
+			resets = <&mcu_ccu RST_BUS_MCU_DMA>;
+			#dma-cells = <1>;
+			status = "okay";
+		};
 	};
 };
+
diff --git a/arch/arm/dts/sun55i-a527-radxa-a5e.dts b/arch/arm/dts/sun55i-a527-radxa-a5e.dts
index ae92905378b..54a4873a11d 100644
--- a/arch/arm/dts/sun55i-a527-radxa-a5e.dts
+++ b/arch/arm/dts/sun55i-a527-radxa-a5e.dts
@@ -6,19 +6,62 @@
 #include "sun55i-a523.dtsi"
 
 #include <dt-bindings/gpio/gpio.h>
+#define USE_PCIE 1
 
 / {
-	model = "Radxa A5E";
-	compatible = "radxa,a5e", "allwinner,sun55i-a527";
+	model = "Radxa Cubie A5E";
+	compatible = "radxa,cubie-a5e", "allwinner,sun55i-a527";
 
 	aliases {
+		ethernet0 = &gmac0;
+		ethernet1 = &gmac1;
+		ethernet2 = &sdio_wifi;
 		serial0 = &uart0;
+		pcie = &pcie;
+		combophy = &combophy;
+		gma340_oe = &gma340_oe;
+		gma340_pcie = &gma340_pcie;
+		reg_pcie_3v3 = &reg_pcie_3v3;
+		tmp_pcie_1v8 = &tmp_pcie_1v8;
 	};
 
 	chosen {
 		stdout-path = "serial0:115200n8";
 	};
 
+	leds {
+		compatible = "gpio-leds";
+
+		led-0 {
+			label = "radxa:green:power";
+			gpios = <&r_pio 0 4 GPIO_ACTIVE_LOW>; /* PL4 */
+			linux,default-trigger = "heartbeat";
+		};
+
+		led-1 {
+			label = "radxa:blue:user";
+			gpios = <&r_pio 0 5 GPIO_ACTIVE_LOW>; /* PL5 */
+			linux,default-trigger = "default-on";
+		};
+	};
+
+	ext_osc32k: ext-osc32k-clk {
+		#clock-cells = <0>;
+		compatible = "fixed-clock";
+		clock-frequency = <32768>;
+		clock-output-names = "ext_osc32k";
+	};
+
+	reg_3v3_wifi: 3v3-wifi {
+		compatible = "regulator-fixed";
+		regulator-name = "3v3-wifi";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_vcc5v>;
+		gpio = <&r_pio 0 7 GPIO_ACTIVE_HIGH>; /* PL7 */
+		enable-active-high;
+	};
+
 	reg_vcc5v: vcc5v {
 		/* board wide 5V supply from the USB-C connector */
 		compatible = "regulator-fixed";
@@ -37,6 +80,64 @@
 		gpio = <&r_pio 0 8 GPIO_ACTIVE_HIGH>;	/* PL8 */
 		enable-active-high;
 	};
+
+	reg_pcie_3v3: pcie-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "pcie-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-enable-ramp-delay = <1000>;
+		regulator-always-on;
+		regulator-boot-on;
+		gpio = <&r_pio 0 11 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	tmp_pcie_1v8: pcie-1v8 {
+		compatible = "regulator-fixed";
+		regulator-name = "pcie-1v8";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		regulator-enable-ramp-delay = <1000>;
+		regulator-always-on;
+		regulator-boot-on;
+	};
+
+	gma340_oe: gma340-oe {
+		compatible = "regulator-fixed";
+		regulator-name = "gma340-oe";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		//regulator-enable-ramp-delay = <1000>;
+		regulator-always-on;
+		regulator-boot-on;
+		gpio = <&pio 1 7 GPIO_ACTIVE_LOW>;
+	};
+
+#ifdef USE_PCIE
+	gma340_pcie: gma340-pcie {
+		compatible = "regulator-fixed";
+		regulator-name = "gma340-pcie";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		//regulator-enable-ramp-delay = <1000>;
+		regulator-always-on;
+		regulator-boot-on;
+		gpio = <&pio 1 6 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+#else
+	gma340_usb3: gma340-usb3 {
+		compatible = "regulator-fixed";
+		regulator-name = "gma340-usb3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		//regulator-enable-ramp-delay = <1000>;
+		regulator-always-on;
+		regulator-boot-on;
+		gpio = <&pio 1 6 GPIO_ACTIVE_LOW>;
+	};
+#endif
 };
 
 &ehci0 {
@@ -47,11 +148,75 @@
 	status = "okay";
 };
 
+&gmac0 {
+	phy-mode = "rgmii-id";
+	phy-handle = <&ext_rgmii0_phy>;
+	phy-supply = <&reg_cldo3>;
+
+	allwinner,tx-delay-ps = <300>;
+	allwinner,rx-delay-ps = <400>;
+
+	status = "okay";
+};
+
+&gmac1 {
+	phy-mode = "rgmii-id";
+	phy-handle = <&ext_rgmii1_phy>;
+	phy-supply = <&reg_cldo4>;
+
+	allwinner,tx-delay-ps = <300>;
+	allwinner,rx-delay-ps = <400>;
+
+	status = "okay";
+};
+
+&gpu {
+	mali-supply = <&reg_dcdc2>;
+	status = "okay";
+};
+
+&mdio0 {
+	ext_rgmii0_phy: ethernet-phy@1 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <1>;
+		reset-gpios = <&pio 7 8 GPIO_ACTIVE_LOW>; /* PH8 */
+		reset-assert-us = <10000>;
+		reset-deassert-us = <150000>;
+	};
+};
+
+&mdio1 {
+	ext_rgmii1_phy: ethernet-phy@1 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <1>;
+		reset-gpios = <&pio 9 16 GPIO_ACTIVE_LOW>; /* PJ16 */
+		reset-assert-us = <10000>;
+		reset-deassert-us = <150000>;
+	};
+};
+
 &mmc0 {
 	vmmc-supply = <&reg_cldo3>;
-	cd-gpios = <&pio 5 6 (GPIO_ACTIVE_LOW | GPIO_PULL_DOWN)>; /* PF6 */
+	cd-gpios = <&pio 5 6 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>; /* PF6 */
+	bus-width = <4>;
+	status = "okay";
+};
+
+&mmc1_pins {
+	drive-strength = <40>;
+};
+
+&mmc1 {
 	bus-width = <4>;
+	vmmc-supply = <&reg_3v3_wifi>;
+	non-removable;
+	// todo: investigate why clock above 40MHz makes data errors
+	max-frequency = <35000000>;
 	status = "okay";
+
+	sdio_wifi: wifi@1 {
+		reg = <1>;
+	};
 };
 
 &ohci0 {
@@ -232,6 +397,9 @@
 			};
 
 			/* DCDC2 is polyphased with DCDC1 */
+			reg_dcdc2_323: dcdc2 {
+				x-powers,polyphased;
+			};
 
 			/* RISC-V management core supply */
 			reg_dcdc3_323: dcdc3 {
@@ -290,3 +458,53 @@
 	usb1_vbus-supply = <&reg_usb_vbus>;
 	status = "okay";
 };
+
+/* PCIE and USB Switch */
+&combophy {
+	resets = <&ccu RST_BUS_PCIE_USB3>;
+#ifdef  USE_PCIE
+	phy_use_sel = <0>; /* 0:PCIE; 1:USB3 */
+#else
+	phy_use_sel = <1>; /* 0:PCIE; 1:USB3 */
+#endif
+	status = "okay";
+};
+
+#ifdef  USE_PCIE
+&pcie {
+	reset-gpios = <&pio 7 11 GPIO_ACTIVE_HIGH>;
+	wake-gpios = <&pio 7 12 GPIO_ACTIVE_HIGH>;
+	num-lanes  = <1>;
+	clk-freq-100M;
+	pcie3v3-supply = <&reg_pcie_3v3>;
+	pcie1v8-supply = <&tmp_pcie_1v8>;
+	status = "okay";
+};
+#else
+&usbc2 {
+	drvvbus-supply = <&reg_usb_vbus>;
+	wakeup-source;
+	usb_detect_mode = <0x0>;
+	aw,vbus-shared-quirk;
+	status = "okay";
+};
+
+&xhci2 {
+	drvvbus-supply = <&reg_usb_vbus>;
+	dr_mode = "host";
+	status = "okay";
+};
+
+&u2phy {
+	status = "okay";
+};
+#endif
+
+&usbc1 {
+	device_type = "usbc1";
+	usb_regulator_io = "nocare";
+	usb_wakeup_suspend = <1>;
+	wakeup-source;
+	status = "okay";
+};
+
diff --git a/include/dt-bindings/clock/sun55i-a523-ccu.h b/include/dt-bindings/clock/sun55i-a523-ccu.h
index c94669156ff..bec8865abf2 100644
--- a/include/dt-bindings/clock/sun55i-a523-ccu.h
+++ b/include/dt-bindings/clock/sun55i-a523-ccu.h
@@ -37,13 +37,13 @@
 #define CLK_PLL_VIDEO2_8X	28
 #define CLK_PLL_VIDEO2_4X	29
 #define CLK_PLL_VIDEO2_3X	30
-#define CLK_PLL_VE		31
-#define CLK_PLL_AUDIO0_4X	32
-#define CLK_PLL_AUDIO0_2X	33
-#define CLK_PLL_AUDIO0		34
-#define CLK_PLL_AUDIO1		35
-#define CLK_PLL_AUDIO1_DIV2	36
-#define CLK_PLL_AUDIO1_DIV5	37
+#define CLK_PLL_VIDEO3_8X	31
+#define CLK_PLL_VIDEO3_4X	32
+#define CLK_PLL_VIDEO3_3X	33
+#define CLK_PLL_VE		34
+#define CLK_PLL_AUDIO0_4X	35
+#define CLK_PLL_AUDIO0_2X	36
+#define CLK_PLL_AUDIO0		37
 #define CLK_PLL_NPU_4X		38
 #define CLK_PLL_NPU_2X		39
 #define CLK_PLL_NPU		40
@@ -83,105 +83,112 @@
 #define CLK_MBUS_DMA		74
 #define CLK_MBUS_VE		75
 #define CLK_MBUS_CE		76
-#define CLK_MBUS_TVIN		77
-#define CLK_MBUS_CSI		78
-#define CLK_BUS_DRAM		79
-#define CLK_NAND0		80
-#define CLK_NAND1		81
-#define CLK_BUS_NAND		82
-#define CLK_MMC0		83
-#define CLK_MMC1		84
-#define CLK_MMC2		85
-#define CLK_BUS_SYSDAP		86
-#define CLK_BUS_MMC0		87
-#define CLK_BUS_MMC1		88
-#define CLK_BUS_MMC2		89
-#define CLK_BUS_UART0		90
-#define CLK_BUS_UART1		91
-#define CLK_BUS_UART2		92
-#define CLK_BUS_UART3		93
-#define CLK_BUS_UART4		94
-#define CLK_BUS_UART5		95
-#define CLK_BUS_UART6		96
-#define CLK_BUS_UART7		97
-#define CLK_BUS_I2C0		98
-#define CLK_BUS_I2C1		99
-#define CLK_BUS_I2C2		100
-#define CLK_BUS_I2C3		101
-#define CLK_BUS_I2C4		102
-#define CLK_BUS_I2C5		103
-#define CLK_BUS_CAN		104
-#define CLK_SPI0		105
-#define CLK_SPI1		106
-#define CLK_SPI2		107
-#define CLK_SPIFC		108
-#define CLK_BUS_SPI0		109
-#define CLK_BUS_SPI1		110
-#define CLK_BUS_SPI2		111
-#define CLK_BUS_SPIFC		112
-#define CLK_EMAC0_25M		113
-#define CLK_EMAC1_25M		114
-#define CLK_BUS_EMAC0		115
-#define CLK_BUS_EMAC1		116
-#define CLK_IR_RX		117
-#define CLK_BUS_IR_RX		118
-#define CLK_IR_TX		119
-#define CLK_BUS_IR_TX		120
-#define CLK_GPADC0		121
-#define CLK_GPADC1		122
-#define CLK_BUS_GPADC0		123
-#define CLK_BUS_GPADC1		124
-#define CLK_BUS_THS		125
-#define CLK_USB_OHCI0		126
-#define CLK_USB_OHCI1		127
-#define CLK_BUS_OHCI0		128
-#define CLK_BUS_OHCI1		129
-#define CLK_BUS_EHCI0		130
-#define CLK_BUS_EHCI1		131
-#define CLK_BUS_OTG		132
-#define CLK_BUS_LRADC		133
-#define CLK_PCIE_AUX		134
-#define CLK_BUS_DPSS_TOP	135
-#define CLK_HDMI_24M		136
-#define CLK_HDMI_CEC_32K	137
-#define CLK_HDMI_CEC		138
-#define CLK_BUS_HDMI		139
-#define CLK_MIPI_DSI0		140
-#define CLK_MIPI_DSI1		141
-#define CLK_BUS_MIPI_DSI0	142
-#define CLK_BUS_MIPI_DSI1	143
-#define CLK_TCON_LCD0		144
-#define CLK_TCON_LCD1		145
-#define CLK_COMBOPHY_DSI0	146
-#define CLK_COMBOPHY_DSI1	147
-#define CLK_BUS_TCON_LCD0	148
-#define CLK_BUS_TCON_LCD1	149
-#define CLK_TCON_TV0		150
-#define CLK_TCON_TV1		151
-#define CLK_BUS_TCON_TV0	152
-#define CLK_BUS_TCON_TV1	153
-#define CLK_EDP			154
-#define CLK_BUS_EDP		155
-#define CLK_LEDC		156
-#define CLK_BUS_LEDC		157
-#define CLK_CSI_TOP		158
-#define CLK_CSI_MCLK0		159
-#define CLK_CSI_MCLK1		160
-#define CLK_CSI_MCLK2		161
-#define CLK_CSI_MCLK3		162
-#define CLK_BUS_CSI		163
-#define CLK_ISP			164
-#define CLK_DSP			165
-#define CLK_BUS_DSP_CFG		166
-#define CLK_FANOUT_24M		167
-#define CLK_FANOUT_12M		168
-#define CLK_FANOUT_16M		169
-#define CLK_FANOUT_25M		170
-#define CLK_FANOUT_32K		171
-#define CLK_FANOUT_27M		172
-#define CLK_FANOUT_PCLK		173
-#define CLK_FANOUT0		174
-#define CLK_FANOUT1		175
-#define CLK_FANOUT2		176
+#define CLK_MBUS_CSI		77
+#define CLK_MBUS_ISP		78
+#define CLK_MBUS_EMAC1		79
+#define CLK_BUS_DRAM		80
+#define CLK_NAND0		81
+#define CLK_NAND1		82
+#define CLK_BUS_NAND		83
+#define CLK_MMC0		84
+#define CLK_MMC1		85
+#define CLK_MMC2		86
+#define CLK_BUS_SYSDAP		87
+#define CLK_BUS_MMC0		88
+#define CLK_BUS_MMC1		89
+#define CLK_BUS_MMC2		90
+#define CLK_BUS_UART0		91
+#define CLK_BUS_UART1		92
+#define CLK_BUS_UART2		93
+#define CLK_BUS_UART3		94
+#define CLK_BUS_UART4		95
+#define CLK_BUS_UART5		96
+#define CLK_BUS_UART6		97
+#define CLK_BUS_UART7		98
+#define CLK_BUS_I2C0		99
+#define CLK_BUS_I2C1		100
+#define CLK_BUS_I2C2		101
+#define CLK_BUS_I2C3		102
+#define CLK_BUS_I2C4		103
+#define CLK_BUS_I2C5		104
+#define CLK_BUS_CAN		105
+#define CLK_SPI0		106
+#define CLK_SPI1		107
+#define CLK_SPI2		108
+#define CLK_SPIFC		109
+#define CLK_BUS_SPI0		110
+#define CLK_BUS_SPI1		111
+#define CLK_BUS_SPI2		112
+#define CLK_BUS_SPIFC		113
+#define CLK_EMAC0_25M		114
+#define CLK_EMAC1_25M		115
+#define CLK_BUS_EMAC0		116
+#define CLK_BUS_EMAC1		117
+#define CLK_IR_RX		118
+#define CLK_BUS_IR_RX		119
+#define CLK_IR_TX		120
+#define CLK_BUS_IR_TX		121
+#define CLK_GPADC0		122
+#define CLK_GPADC1		123
+#define CLK_BUS_GPADC0		124
+#define CLK_BUS_GPADC1		125
+#define CLK_BUS_THS		126
+#define CLK_USB_OHCI0		127
+#define CLK_USB_OHCI1		128
+#define CLK_BUS_OHCI0		129
+#define CLK_BUS_OHCI1		130
+#define CLK_BUS_EHCI0		131
+#define CLK_BUS_EHCI1		132
+#define CLK_BUS_OTG		133
+#define CLK_BUS_LRADC		134
+#define CLK_PCIE_AUX		135
+#define CLK_BUS_DISPLAY0_TOP	136
+#define CLK_BUS_DISPLAY1_TOP	137
+#define CLK_HDMI_24M		138
+#define CLK_HDMI_CEC_32K	139
+#define CLK_HDMI_CEC		140
+#define CLK_BUS_HDMI		141
+#define CLK_MIPI_DSI0		142
+#define CLK_MIPI_DSI1		143
+#define CLK_BUS_MIPI_DSI0	144
+#define CLK_BUS_MIPI_DSI1	145
+#define CLK_TCON_LCD0		146
+#define CLK_TCON_LCD1		147
+#define CLK_TCON_LCD2		148
+#define CLK_COMBOPHY_DSI0	149
+#define CLK_COMBOPHY_DSI1	150
+#define CLK_BUS_TCON_LCD0	151
+#define CLK_BUS_TCON_LCD1	152
+#define CLK_BUS_TCON_LCD2	153
+#define CLK_TCON_TV0		154
+#define CLK_TCON_TV1		155
+#define CLK_BUS_TCON_TV0	156
+#define CLK_BUS_TCON_TV1	157
+#define CLK_EDP			158
+#define CLK_BUS_EDP		159
+#define CLK_LEDC		160
+#define CLK_BUS_LEDC		161
+#define CLK_CSI_TOP		162
+#define CLK_CSI_MCLK0		163
+#define CLK_CSI_MCLK1		164
+#define CLK_CSI_MCLK2		165
+#define CLK_CSI_MCLK3		166
+#define CLK_BUS_CSI		167
+#define CLK_ISP			168
+#define CLK_DSP			169
+#define CLK_FANOUT_24M		170
+#define CLK_FANOUT_12M		171
+#define CLK_FANOUT_16M		172
+#define CLK_FANOUT_25M		173
+#define CLK_FANOUT_27M		174
+#define CLK_FANOUT_PCLK		175
+#define CLK_FANOUT0		176
+#define CLK_FANOUT1		177
+#define CLK_FANOUT2		178
+#define CLK_NPU			179
+#define CLK_USB3_MBUS_GATE	180
+#define CLK_USB3_REF		181
+#define CLK_USB2_REF		182
+#define CLK_USB3_SUSPEND	183
 
 #endif /* _DT_BINDINGS_CLK_SUN55I_A523_CCU_H_ */
diff --git a/include/dt-bindings/reset/sun55i-a523-r-ccu.h b/include/dt-bindings/reset/sun55i-a523-r-ccu.h
index dd6fbb372e1..eb31ae9958d 100644
--- a/include/dt-bindings/reset/sun55i-a523-r-ccu.h
+++ b/include/dt-bindings/reset/sun55i-a523-r-ccu.h
@@ -21,5 +21,6 @@
 #define RST_BUS_R_IR_RX		12
 #define RST_BUS_R_RTC		13
 #define RST_BUS_R_CPUCFG	14
+#define RST_BUS_R_PPU0		15
 
 #endif /* _DT_BINDINGS_RST_SUN55I_A523_R_CCU_H_ */